src/main/java/codeu/model/store/basic/IndexStore.java

```diff
@@ -66,13 +71,96 @@ public void addMapping (String word, Message message) {
   /**
     * Search function for a single word, returns a Set of messages from the Hashtable if the word is in it, otherwise return null
     */
-  public Set<Message> searchWord (String word) {
-    if (wordsHash.containsKey(word)) {
-      return wordsHash.get(word);
+  public List<Message> search (String searchTarget) {
+    if (searchTarget.contains("||")) {
+      return searchUnion(searchTarget);
+    }else if (searchTarget.contains("&&")) {
+      return searchIntersection(searchTarget);
+    }else if (wordsHash.containsKey(searchTarget)){
+      List<Message> messageList = new ArrayList<Message>(wordsHash.get(searchTarget));
+      return sortMessageList(messageList);
     }
     return null;
   }
 
+  /**
+    * Pre: the given string contains "||"
+    * returns the union of the elements in the given searchTarget
+    */
+  public List<Message> searchUnion (String searchTarget) {
+    // remove bad format || in searchtarget
+    if (searchTarget.indexOf("|") == 0) {
+      searchTarget = searchTarget.substring(2);
+    }
+    if (searchTarget.indexOf("||") == (searchTarget.length()-2)) {
+      searchTarget = searchTarget.substring(0,searchTarget.length()-2);
+    }
+    String[] words = searchTarget.split("\\|\\|");
+
+    if (words.length == 0) {
+      return null;
+    }else {
+      Set<Message> hashSetResult = new HashSet<Message>(search(words[0]));
```

Comment:
Does anything bad here happen if search returns null?

Reply:
right i think it gives a null pointer exception if it's null. I think a better solution would be to create an empty one first and add things to it in the for loop under, this prevents null results in initializing and also takes care of the redundant call mentioned under